File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/C.Странное_сравнение/solution.cpp
#include <iostream>
#include <string>


void solution(std::string& a, std::string& b) {
    /*
    Time Complexity O(N)
    Memory Complexity O(1)
    */
    bool isValid = true;
    int aLeft = 0, bLeft = 0;
    int aRight = 0, bRight = 0;
    while (aRight < a.size() && bRight < b.size()) {
        while (aRight < a.size() && a[aLeft] == a[aRight]) {
            ++aRight;
        }
        while (bRight < b.size() && b[bLeft] == b[bRight]) {
            ++bRight;
        }
        if (aRight - aLeft != bRight - bLeft) {
            isValid = false;
            break;
        } else {
            aLeft = aRight;
            bLeft = bRight;
        }
    }
    std::cout << (isValid && aRight == a.size() && bRight == b.size()? "YES" : "NO") << "\n";
}


int main() {
    std::string a, b;
    std::cin >> a >> b;
    solution(a, b);
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/C.Странное_сравнение/solution.py
import sys


def solution(a: str, b: str) -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(1)
    """
    is_valid: bool = True
    a_left: int = 0
    a_right: int = 0
    b_left: int = 0
    b_right: int = 0
    while a_right < len(a) and b_right < len(b):
        while a_right < len(a) and a[a_left] == a[a_right]:
            a_right += 1
        while b_right < len(b) and b[b_left] == b[b_right]:
            b_right += 1
        if a_right - a_left != b_right - b_left:
            is_valid = False
            break
        else:
            a_left = a_right
            b_left = b_right
    sys.stdout.write(['NO', 'YES'][is_valid and a_right == len(a) and b_right == len(b)] + '\n')


if __name__ == '__main__':
    a: str = sys.stdin.readline().rstrip()
    b: str = sys.stdin.readline().rstrip()
    solution(a, b)
